libsolidity/interface/CompilerStack.cpp

```diff
@@ -103,6 +107,98 @@ CompilerStack::~CompilerStack()
 	TypeProvider::reset();
 }
 
+void CompilerStack::createAndAssignCallGraphs()
+{
+	for (Source const* source: m_sourceOrder)
+		if (source->ast)
+			for (ASTPointer<ASTNode> const& node: source->ast->nodes())
+				if (auto const* contract = dynamic_cast<ContractDefinition*>(node.get()))
+				{
+					Contract& contractState = m_contracts.at(contract->fullyQualifiedName());
+
+					contractState.contract->annotation().creationCallGraph = make_unique<CallGraph>(
+						FunctionCallGraphBuilder::buildCreationGraph(
+							*contract
+						)
+					);
+					contractState.contract->annotation().deployedCallGraph = make_unique<CallGraph>(
+						FunctionCallGraphBuilder::buildDeployedGraph(
+							*contract,
+							**contractState.contract->annotation().creationCallGraph
+						)
+					);
+
+					auto contractDeps = ranges::views::concat(
+						contract->annotation().creationCallGraph->get()->bytecodeDependency,
+						contract->annotation().deployedCallGraph->get()->bytecodeDependency
+					);
+
+					for (auto const& [dependencyContract, referencee]: contractDeps)
+					{
+						solAssert(dependencyContract != nullptr, "");
+						contractState.contract->annotation().contractDependencies.emplace(dependencyContract, referencee);
+					}
+				}
+}
+
+void CompilerStack::findAndReportCyclicContractDependencies()
+{
+	// Cycles we found, used to avoid duplicate reports for the same reference
+	map<ASTNode const*, ContractDefinition const*, ASTNode::CompareByID> foundCycles;
+
+	// Use an extra struct to remember the reference for the found cycle
+	struct RememberingCycleDetector : public util::CycleDetector<ContractDefinition>
+	{
+		RememberingCycleDetector(Visitor _visit):
+			util::CycleDetector<ContractDefinition>(_visit)
+		{}
+
+		ASTNode const* m_cycleReferencee = nullptr;
+	} cycleDetector{[&](ContractDefinition const& _contract, util::CycleDetector<ContractDefinition>& _cycleDetector, size_t _depth)
+	{
+		if (_depth >= MaxCycleDetectionRecursionDepth)
+			m_errorReporter.fatalTypeError(7864_error, _contract.location(), "Contract dependencies exhausting cyclic dependency validator");
+
+		for (auto& [dependencyContract, referencee]: _contract.annotation().contractDependencies)
+			if (ContractDefinition const* cycle = _cycleDetector.run(*dependencyContract))
+			{
+				if (!cycleDetector.m_cycleReferencee)
+				{
+					cycleDetector.m_cycleReferencee = _contract.annotation().contractDependencies.at(cycle);
+					solAssert(cycleDetector.m_cycleReferencee, "");
+				}
+				return;
+			}
+	}};
+
+	for (Source const* source: m_sourceOrder)
+		if (source->ast)
+			for (ASTPointer<ASTNode> const& node: source->ast->nodes())
+				if (auto const* contractDefinition = dynamic_cast<ContractDefinition*>(node.get()))
+				{
+					ContractDefinition const* cycle = cycleDetector.run(*contractDefinition);
+
+					if (cycle)
+					{
+						solAssert(cycleDetector.m_cycleReferencee, "");
+						foundCycles.emplace(cycleDetector.m_cycleReferencee, cycle);
```

Comment:
Why don't you report it right here?

Reply:
You can use a simple set to track what you already reported, but you don't need to store the other information.